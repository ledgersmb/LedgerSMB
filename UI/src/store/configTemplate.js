/** @format */

export const configStoreTemplate = {
    // to be mixed in at the parent level:
    // state: () => {
    //    return {
    //        "fields": ["id", "description"],
    //        "items": [],
    //        "_links": [],
    //        "url": "products/warehouses/"
    //    };
    // },
    getters: {
        apiURL: (state) => `/erp/api/v0/${state.url}`
    },
    actions: {
        async initialize() {
            const response = await fetch(`/erp/api/v0/${this.url}`, {
                method: "GET"
            });

            if (response.ok) {
                const rv = await response.json();
                this.items = rv.items;
                this._links = rv._links;
            } else {
                throw new Error(`HTTP Error: ${response.status}`);
            }
        },
        async add(adding) {
            const response = await fetch(`/erp/api/v0/${this.url}`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json"
                },
                body: JSON.stringify(adding)
            });

            if (response.ok) {
                const added = await response.json();
                added._meta = {
                    ETag: response.headers.get("ETag")
                };
                this.items.push(added);
            } else {
                throw new Error(`HTTP Error: ${response.status}`);
            }
        },
        async del(id) {
            const warehouse = this.getById(id);
            const response = await fetch(`/erp/api/v0/${this.url}/${id}`, {
                method: "DELETE",
                headers: {
                    "If-Match": warehouse._meta.ETag
                }
            });

            if (response.ok) {
                let index = this.items.findIndex((w) => w[this.id] === id);
                if (index !== -1) {
                    this.items.splice(index, 1);
                }
            } else {
                throw new Error(`HTTP Error: ${response.status}`);
            }
        },
        async get(id) {
            let index = this.items.findIndex((w) => w[this.id] === id);
            if (index === -1) {
                index = this.items.length;
            }
            const warehouse = this.items[index];
            if (!warehouse || !warehouse._meta || warehouse._meta.invalidated) {
                const response = await fetch(`/erp/api/v0/${this.url}/${id}`, {
                    method: "GET"
                });

                if (response.ok) {
                    const newData = await response.json();
                    newData[this.id] = id;
                    newData._meta = {
                        ETag: response.headers.get("ETag")
                    };
                    this.items[index] = newData;
                } else {
                    throw new Error(`HTTP Error: ${response.status}`);
                }
            }

            return this.items[index];
        },
        getById(id) {
            if (id === "") {
                const rv = {};
                this.fields.forEach((f) => {
                    rv[f] = undefined;
                });
                return rv;
            }
            return this.items.find((w) => w[this.id] === id);
        },
        async save(id, data) {
            const warehouse = this.getById(id);
            const response = await fetch(`/erp/api/v0/${this.url}/${id}`, {
                method: "PUT",
                headers: {
                    "Content-Type": "application/json",
                    "If-Match": warehouse._meta.ETag
                },
                body: JSON.stringify(data)
            });

            if (response.ok) {
                const newData = await response.json();
                const index = this.items.findIndex((w) => w[this.id] === id);
                newData[this.id] = id; // prevent overwriting 'id'
                newData._meta = {
                    ETag: response.headers.get("ETag")
                };
                this.items[index] = newData;
            } else {
                throw new Error(`HTTP Error: ${response.status}`);
            }
        }
    }
};
