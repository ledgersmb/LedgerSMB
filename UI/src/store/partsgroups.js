/** @format */

import { defineStore } from "pinia";

export const usePartsgroupsStore = defineStore("partsgroups", {
    actions: {
        async initialize() {
            const response = await fetch(`./erp/api/v0/${this.url}`, {
                method: "GET"
            });

            if (response.ok) {
                const rv = await response.json();
                this.items = rv.items;
                this._links = rv._links;
            } else {
                throw new Error(`HTTP Error: ${response.status}`);
            }
        },
        async add(adding) {
            const response = await fetch(`./erp/api/v0/${this.url}`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json"
                },
                body: JSON.stringify(adding)
            });

            if (response.ok) {
                const added = await response.json();
                added._meta = {
                    ETag: response.headers.get("ETag")
                };
                this.items.push(added);
            } else {
                throw new Error(`HTTP Error: ${response.status}`);
            }
        },
        async del(id) {
            const partsgroup = this.getById(id);
            const response = await fetch(`./erp/api/v0/${this.url}/${id}`, {
                method: "DELETE",
                headers: {
                    "If-Match": `"${partsgroup._meta.ETag}"`
                }
            });

            if (response.ok) {
                let index = this.items.findIndex((w) => w.id === id);
                if (index !== -1) {
                    this.items.splice(index, 1);
                }
            } else {
                throw new Error(`HTTP Error: ${response.status}`);
            }
        },
        getById(id) {
            return this.items.find((w) => w.id === id);
        },
        async save(id, data) {
            const partsgroup = this.getById(id);
            const response = await fetch(`./erp/api/v0/${this.url}/${id}`, {
                method: "PUT",
                headers: {
                    "Content-Type": "application/json",
                    "If-Match": `"${partsgroup._meta.ETag}"`
                },
                body: JSON.stringify(data)
            });

            if (response.ok) {
                const newData = await response.json();
                const index = this.items.findIndex((w) => w.id === id);
                newData.id = id; // prevent overwriting 'id'
                newData._meta = {
                    ETag: response.headers.get("ETag")
                };
                this.items[index] = newData;
            } else {
                throw new Error(`HTTP Error: ${response.status}`);
            }
        }
    },
    getters: {
        apiURL: (state) => `/erp/api/v0/${state.url}`,
        tree: (state) => {
            let byId = {};
            state.items.forEach((i) => {
                byId[i.id] = i;
                i.children = [];
            });
            state.items.forEach((i) => {
                if (i.parent) {
                    byId[i.parent].children.push(i);
                }
            });
            return state.items.filter((i) => !i.parent);
        }
    },
    state: () => {
        return {
            items: [],
            _links: [],
            url: "products/partsgroups"
        };
    }
});
