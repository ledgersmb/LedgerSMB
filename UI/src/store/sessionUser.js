/** @format */

import { defineStore } from "pinia";

export const useSessionUserStore = defineStore("sessionUser", {
    state: () => {
        return {
            roles: [],
            preferences: {}
        };
    },
    actions: {
        async initialize() {
            const response = await fetch("./erp/api/v0/session", {
                method: "GET"
            });

            if (response.ok) {
                let data = await response.json();
                this.$patch({
                    roles: data.roles,
                    preferences: data.preferences
                });
            } else {
                throw new Error(`HTTP Error: ${response.status}`);
            }
        }
    },
    getters: {
        hasRole: (state) => {
            return (role) => state.roles.find((r) => r === role) !== undefined;
        }
    }
});
