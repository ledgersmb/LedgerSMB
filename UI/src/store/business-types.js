/** @format */

import { defineStore } from "pinia";
import { configStoreTemplate } from "@/store/configTemplate";

export const useBusinessTypesStore = defineStore("business-types", {
    ...configStoreTemplate,
    state: () => {
        return {
            fields: ["id", "description", "discount"],
            id: "id",
            items: [],
            url: "contacts/business-types"
        };
    }
});
