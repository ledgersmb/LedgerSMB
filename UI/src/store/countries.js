/** @format */

import { defineStore } from "pinia";
import { configStoreTemplate } from "@/store/configTemplate";

let { actions: configActions, getters: configGetters } = configStoreTemplate;

export const useCountriesStore = defineStore("countries", {
    ...configStoreTemplate,
    getters: {
        ...configGetters,
        default: (state) => state.items.find((elm) => elm.default)
    },
    actions: {
        ...configActions,
        async setDefault(id) {
            const oldDefault = this.default;
            if (oldDefault && oldDefault[this.id] === id) {
                return;
            }

            const country = this.getById(id);
            await this.save(id, { ...country, default: true });
            if (oldDefault) {
                oldDefault.default = false; // remove the old default
            }
        }
    },
    state: () => {
        return {
            fields: ["_meta", "code", "default", "name"],
            id: "code",
            items: [],
            _links: [],
            url: "countries"
        };
    }
});
