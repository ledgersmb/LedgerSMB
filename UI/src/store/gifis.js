/** @format */

import { defineStore } from "pinia";
import { configStoreTemplate } from "@/store/configTemplate";

export const useGIFIsStore = defineStore("gifis", {
    ...configStoreTemplate,
    state: () => {
        return {
            fields: ["accno", "description"],
            id: "accno",
            items: [],
            url: "gl/gifi"
        };
    }
});
