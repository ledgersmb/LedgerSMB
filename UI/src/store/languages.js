/** @format */

import { defineStore } from "pinia";
import { configStoreTemplate } from "@/store/configTemplate";

let { actions: configActions, getters: configGetters } = configStoreTemplate;

export const useLanguagesStore = defineStore("languages", {
    ...configStoreTemplate,
    getters: {
        ...configGetters,
        default: (state) => state.items.find((elm) => elm.default)
    },
    actions: {
        ...configActions,
        async setDefault(id) {
            const oldDefault = this.default;
            if (oldDefault && oldDefault[this.id] === id) {
                return;
            }

            const language = this.getById(id);
            await this.save(id, { ...language, default: true });
            if (oldDefault) {
                oldDefault.default = false; // remove the old default
            }
        }
    },
    state: () => {
        return {
            fields: ["_meta", "code", "default", "description"],
            id: "code",
            items: [],
            _links: [],
            url: "languages"
        };
    }
});
