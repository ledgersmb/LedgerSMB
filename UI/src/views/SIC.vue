<!-- @format -->
<!-- eslint-disable prettier/prettier -->

<script setup>
import { useSICsStore } from "@/store/sics";
import { useI18n } from "vue-i18n";
import ConfigTable from "@/components/ConfigTable.vue";

import ImportCsvSic from "@/views/ImportCSV-SIC";

const { t } = useI18n();
const COLUMNS = [
    { key: "code",        type: "text",     head: t("Code") },
    { key: "description", type: "text",     head: t("Description") },
];

const store = useSICsStore();

</script>

<template>
    <h1 class="listtop">{{ t("Configure Standard Industry Codes (SIC)") }}</h1>
    <ConfigTable
        :columns="COLUMNS"
        :store="store"
        storeId="code"
        editRole="sic_edit"
        createRole="sic_create"
    />
    <div class="import-section">
        <h2 class="listheading">{{ t("Import") }}</h2>
        <ImportCsvSic @upload-success="store.initialize()" />
    </div>
</template>

<style scoped>
.import-section {
    margin-top: 3em;
}

.import-section h2 {
    text-align: left;
}
</style>
