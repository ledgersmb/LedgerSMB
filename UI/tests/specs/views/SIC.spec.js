/** @format */
/* global retry */

import SIC from "@/views/SIC.vue";
import { useSessionUserStore } from "@/store/sessionUser";
import { factory } from "./factory";

let wrapper;
let sessionUser;

describe("SIC - register as a component", () => {
    beforeEach(() => {
        wrapper = factory(SIC);
        sessionUser = useSessionUserStore();
    });
    afterEach(() => {
        // wrapper.unmount();
    });

    it("should show dialog", async () => {
        expect(wrapper.exists()).toBeTruthy();

        // Check loading
        expect(wrapper.get(".dynatableLoading").text()).toBe("Loading...");

        // Wait until loading done
        await retry(() =>
            expect(wrapper.find(".dynatableData").isVisible()).toBe(true)
        );

        // Validate against snapshot
        expect(wrapper.element).toMatchSnapshot();

        const sic_items = wrapper.findAll(".data-row");
        expect(sic_items).toHaveLength(2);

        // Validate data displayed
        let data = sic_items.map((rows) => {
            return rows.findAll(".input-box").map((row) => row.element.value);
        });
        expect(data).toEqual([
            ["541330", "Engineering service"],
            ["611430", "Professional and management development training"]
        ]);

        // TODO: Test links
        // expect that the links displayed match
        // what was returned by the API
    });

    it("should show dialog with editable data", async () => {
        // Give user edition capability
        sessionUser.$patch({ roles: ["sic_edit"] });

        expect(wrapper.exists()).toBeTruthy();

        // Check loading
        expect(wrapper.get(".dynatableLoading").text()).toBe("Loading...");

        // Wait until loading done
        await retry(() =>
            expect(wrapper.find(".dynatableData").isVisible()).toBe(true)
        );

        // Validate against snapshot
        expect(wrapper.element).toMatchSnapshot();

        const sic_items = wrapper.findAll(".data-row");
        expect(sic_items).toHaveLength(2);

        // Validate data displayed
        let data = sic_items.map((rows) => {
            return rows.findAll(".input-box").map((row) => row.element.value);
        });
        expect(data).toEqual([
            ["541330", "Engineering service"],
            ["611430", "Professional and management development training"]
        ]);

        // Validate the buttons
        const buttons = sic_items.map((rows) => {
            return rows.findAll("button").map((row) => row.element.name);
        });
        expect(buttons).toEqual([
            ["modify", "save", "cancel"],
            ["modify", "save", "cancel"]
        ]);
    });
});
