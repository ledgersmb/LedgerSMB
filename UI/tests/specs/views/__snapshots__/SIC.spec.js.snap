// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SIC - register as a component should show dialog 1`] = `
<div
  data-v-app=""
>
  
  <h1
    class="listtop"
  >
    Configure Standard Industry Codes (SIC)
  </h1>
  
  <div>
    <table
      class="dynatable"
    >
      <thead>
        <tr>
          <!--v-if-->
          
          <th>
            Code
          </th>
          <th>
            Description
          </th>
          
          <th />
        </tr>
      </thead>
      
      <tbody
        class="dynatableData"
      >
        
        
        <!-- eslint-disable prettier/prettier -->
        <tr
          class="data-row"
        >
          <!--v-if-->
          
          <td
            class="data-entry"
          >
            <input
              class="neutral input-box"
              name="code"
              readonly=""
              type="text"
              value="541330"
            />
          </td>
          <td
            class="data-entry"
          >
            <input
              class="neutral input-box"
              name="description"
              readonly=""
              type="text"
              value="Engineering service"
            />
          </td>
          
          <td>
            <!--v-if-->
            <!--v-if-->
          </td>
        </tr>
        
        
        <!-- eslint-disable prettier/prettier -->
        <tr
          class="data-row"
        >
          <!--v-if-->
          
          <td
            class="data-entry"
          >
            <input
              class="neutral input-box"
              name="code"
              readonly=""
              type="text"
              value="611430"
            />
          </td>
          <td
            class="data-entry"
          >
            <input
              class="neutral input-box"
              name="description"
              readonly=""
              type="text"
              value="Professional and management development training"
            />
          </td>
          
          <td>
            <!--v-if-->
            <!--v-if-->
          </td>
        </tr>
        
        
      </tbody>
      <!--v-if-->
      
    </table>
  </div>
  <div>
    <span>
      Download: 
    </span>
    
    <span>
      <a
        href="http://localhost/erp/api/v0/contacts/sic?format=HTML"
        target="_blank"
      >
        [ HTML ]
      </a>
    </span>
    
  </div>
  
  <div
    class="import-section"
  >
    <h2
      class="listheading"
    >
      Import
    </h2>
    <div
      id="import-sic"
    >
      
      <!-- eslint-disable prettier/prettier -->
      <div
        id="sic"
      >
        <!--v-if-->
        <form
          name="csvupload"
        >
          <input
            name="type"
            type="hidden"
            value="sic"
          />
          <!--v-if-->
          <div
            class="non-grid"
          >
            <!--v-if-->
            <div>
              <label>
                From File
              </label>
              <span>
                <input
                  accept=".csv"
                  name="import_file"
                  required=""
                  type="file"
                />
              </span>
            </div>
          </div>
          <div
            class="inputrow non-grid"
            id="buttonrow"
          >
            <input
              name="trans_type"
              type="hidden"
            />
            <button
              class="submit"
              disabled=""
              name="action"
            >
              
              Save
              
            </button>
          </div>
          <div
            class="details"
          >
            
             The following fields are expected (in this order): 
            <dl>
              <dt>
                code
              </dt>
              <dd />
              <dt>
                sictype
              </dt>
              <dd />
              <dt>
                description
              </dt>
              <dd />
            </dl>
            
          </div>
        </form>
      </div>
      
    </div>
  </div>
  
</div>
`;

exports[`SIC - register as a component should show dialog with editable data 1`] = `
<div
  data-v-app=""
>
  
  <h1
    class="listtop"
  >
    Configure Standard Industry Codes (SIC)
  </h1>
  
  <div>
    <table
      class="dynatable"
    >
      <thead>
        <tr>
          <!--v-if-->
          
          <th>
            Code
          </th>
          <th>
            Description
          </th>
          
          <th />
        </tr>
      </thead>
      
      <tbody
        class="dynatableData"
      >
        
        
        <!-- eslint-disable prettier/prettier -->
        <tr
          class="data-row"
        >
          <!--v-if-->
          
          <td
            class="data-entry"
          >
            <input
              class="neutral input-box"
              name="code"
              readonly=""
              type="text"
              value="541330"
            />
          </td>
          <td
            class="data-entry"
          >
            <input
              class="neutral input-box"
              name="description"
              readonly=""
              type="text"
              value="Engineering service"
            />
          </td>
          
          <td>
            
            <button
              name="modify"
            >
              
              Modify
              
            </button>
            <button
              disabled=""
              name="save"
            >
              
              Save
              
            </button>
            <button
              disabled=""
              name="cancel"
            >
              
              Cancel
              
            </button>
            <!--v-if-->
            
            <!--v-if-->
          </td>
        </tr>
        
        
        <!-- eslint-disable prettier/prettier -->
        <tr
          class="data-row"
        >
          <!--v-if-->
          
          <td
            class="data-entry"
          >
            <input
              class="neutral input-box"
              name="code"
              readonly=""
              type="text"
              value="611430"
            />
          </td>
          <td
            class="data-entry"
          >
            <input
              class="neutral input-box"
              name="description"
              readonly=""
              type="text"
              value="Professional and management development training"
            />
          </td>
          
          <td>
            
            <button
              name="modify"
            >
              
              Modify
              
            </button>
            <button
              disabled=""
              name="save"
            >
              
              Save
              
            </button>
            <button
              disabled=""
              name="cancel"
            >
              
              Cancel
              
            </button>
            <!--v-if-->
            
            <!--v-if-->
          </td>
        </tr>
        
        
      </tbody>
      <!--v-if-->
      
    </table>
  </div>
  <div>
    <span>
      Download: 
    </span>
    
    <span>
      <a
        href="http://localhost/erp/api/v0/contacts/sic?format=HTML"
        target="_blank"
      >
        [ HTML ]
      </a>
    </span>
    
  </div>
  
  <div
    class="import-section"
  >
    <h2
      class="listheading"
    >
      Import
    </h2>
    <div
      id="import-sic"
    >
      
      <!-- eslint-disable prettier/prettier -->
      <div
        id="sic"
      >
        <!--v-if-->
        <form
          name="csvupload"
        >
          <input
            name="type"
            type="hidden"
            value="sic"
          />
          <!--v-if-->
          <div
            class="non-grid"
          >
            <!--v-if-->
            <div>
              <label>
                From File
              </label>
              <span>
                <input
                  accept=".csv"
                  name="import_file"
                  required=""
                  type="file"
                />
              </span>
            </div>
          </div>
          <div
            class="inputrow non-grid"
            id="buttonrow"
          >
            <input
              name="trans_type"
              type="hidden"
            />
            <button
              class="submit"
              disabled=""
              name="action"
            >
              
              Save
              
            </button>
          </div>
          <div
            class="details"
          >
            
             The following fields are expected (in this order): 
            <dl>
              <dt>
                code
              </dt>
              <dd />
              <dt>
                sictype
              </dt>
              <dd />
              <dt>
                description
              </dt>
              <dd />
            </dl>
            
          </div>
        </form>
      </div>
      
    </div>
  </div>
  
</div>
`;
